internal/prefixcollector/collector_test.go

```diff
@@ -0,0 +1,196 @@
+	"k8s.io/client-go/kubernetes/fake"
+
+	"cmd-exclude-prefixes-k8s/internal/utils"
+	"io/ioutil"
+	"testing"
+)
+
+const (
+	configMapPath            = "./testfiles/configMap.yaml"
+	kubeConfigMapPath        = "./testfiles/kubeAdmConfigMap.yaml"
+	testExcludedPrefixesPath = "./testfiles/excludedPrefixes.yaml"
+	configMapName            = "test"
+)
+
+type dummyPrefixSource struct {
+	prefixes []string
+}
+
+type dummyListener struct {
+	notifyChan chan struct{}
+}
+
+func (d *dummyPrefixSource) Start(notifyChan chan<- struct{}) {
+	go func() { utils.Notify(notifyChan) }()
+}
+
+func (d *dummyPrefixSource) Prefixes() []string {
+	return d.prefixes
+}
+
+func (dl *dummyListener) Notify() {
+	utils.Notify(dl.notifyChan)
+}
+
+func newDummyPrefixSource(prefixes []string) *dummyPrefixSource {
+	return &dummyPrefixSource{prefixes}
+}
+
+func newDummyListener() *dummyListener {
+	return &dummyListener{
+		notifyChan: make(chan struct{}, 1),
+	}
+}
+
+func TestCollectorWithDummySources(t *testing.T) {
+	sources := []prefixcollector.ExcludePrefixSource{
+		newDummyPrefixSource(
+			[]string{
+				"127.0.0.1/16",
+				"127.0.2.1/16",
+				"168.92.0.1/24",
+			},
+		),
+		newDummyPrefixSource(
+			[]string{
+				"127.0.3.1/18",
+				"134.56.0.1/8",
+				"168.92.0.1/16",
+			},
+		),
+	}
+
+	expectedResult := []string{
+		"127.0.0.0/16",
+		"168.92.0.0/16",
+		"134.0.0.0/8",
+	}
+
+	testCollector(t, expectedResult, sources)
+}
+
+func TestKubeAdmConfigSource(t *testing.T) {
+	expectedResult := []string{
+		"10.244.0.0/16",
+		"10.96.0.0/12",
+	}
+
+	ctx := createConfigMap(t, prefixcollector.KubeNamespace, kubeConfigMapPath)
+	sources := []prefixcollector.ExcludePrefixSource{
+		prefixcollector.NewKubeAdmPrefixSource(ctx),
+	}
+
+	testCollector(t, expectedResult, sources)
+}
+
+func TestConfigMapSource(t *testing.T) {
+	expectedResult := []string{
+		"168.0.0.0/10",
+		"1.0.0.0/11",
+	}
+
+	ctx := createConfigMap(t, prefixcollector.DefaultConfigMapNamespace, configMapPath)
+	sources := []prefixcollector.ExcludePrefixSource{
+		prefixcollector.NewConfigMapPrefixSource(ctx, configMapName, prefixcollector.DefaultConfigMapNamespace),
+	}
+
+	testCollector(t, expectedResult, sources)
+}
+
+func createConfigMap(t *testing.T, namespace, configPath string) context.Context {
+	ctx := context.Background()
+	clientSet := fake.NewSimpleClientset()
+	configMap := getConfigMap(t, configPath)
+
+	ctx = context.WithValue(ctx, utils.ClientSetKey, clientSet)
+	_, _ = clientSet.CoreV1().
+		ConfigMaps(namespace).
+		Create(ctx, configMap, metav1.CreateOptions{})
+
+	return ctx
+}
+
+func testCollector(t *testing.T, expectedResult []string, sources []prefixcollector.ExcludePrefixSource) {
+	options := []prefixcollector.ExcludePrefixCollectorOption{
+		prefixcollector.WithSources(sources),
+		prefixcollector.WithFilePath(testExcludedPrefixesPath),
+	}
+
+	collector := prefixcollector.NewExcludePrefixCollector(context.Background(), options...)
+	listener := newDummyListener()
```

Comment:
You also can pass notifychan via WithNotifyChan and have access to the chan in the test. I think listeners apporoch is redundant.
```suggestion
        ch := make(chan struct{})
	collector := prefixcollector.NewExcludePrefixCollector(context.Background(), WithNotifyChan(ch), options...)
	// use ch ...
```

Reply:
But we will have race condition while waiting for notification from source?

Reply:
In this case, we can delete WithNotifyChan option too. 
Can we just use `require.Eventually` and monitor file changes in the test?

Reply:
Yes. I added filewatcher and `require.ElementsMatch` after file monitoring.